server/services/invidious.ts

```diff
@@ -0,0 +1,382 @@
+			const [dashProbe, hlsProbe] = await Promise.all([
+				this.probeManifest(host, id, "dash"),
+				this.probeManifest(host, id, "hls"),
+			]);
+			const pick =
+				dashProbe && hlsProbe
+					? dashProbe.topKbps >= hlsProbe.topKbps
+						? dashProbe
+						: hlsProbe
+					: dashProbe ?? hlsProbe;
+			if (pick) {
+				log.debug("Picked streaming manifest", {
+					kind: pick.kind,
+					url: pick.url,
+					topKbps: pick.topKbps,
+					topRes: pick.topRes,
+				});
+				if (pick.kind === "dash") {
+					return {
+						service: "dash",
+						id: pick.url,
+						...base,
+						dash_url: pick.url,
+						mime: "application/dash+xml", // Some downstream codepaths expect a MIME hint even though dash.js can infer. – Pipeline/browser integration
+					};
+				} else {
+					return {
+						service: "hls",
+						id: pick.url,
+						...base,
+						hls_url: pick.url,
+						// Mimetype application/vnd.apple.mpegurl not supported by plyrplayer so switching to x-mpegURL
+						mime: "application/x-mpegURL",
+					};
+				}
+			}
+		} catch (e) {
+			log.warn(`DASH/HLS probing failed for ${host}:${id}, will try progressive.`, {
+				err: e instanceof Error ? e.message : e,
+			});
+		}
+
+		// 2) Progressive fallback (MP4/WebM with audio) via /latest_version
+		// Direct YouTube progressive URLs are short-lived and CORS-restricted; proxying via Invidious stabilizes access for browsers. – YouTube+CORS
+		// Helper to extract numeric quality from labels like "720p", "480p", etc.
+		const q = (label?: string) => {
+			const m = (label || "").match(/(\d+)/);
+			return m ? parseInt(m[1], 10) : 0;
+		};
+
+		const list = (inv.formatStreams || []).filter(f => !!f.url);
+		if (list.length) {
+			const mp4 = list.filter(
+				f => f.container === "mp4" || (f.type || "").toLowerCase().includes("mp4")
+			);
+
+			const score = (f: InvidiousFormat) => q(f.qualityLabel) * 1_000_000 + (f.bitrate || 0);
+			const bestMp4 = maxBy(mp4, score);
+			const bestAny = maxBy(list, score)!;
+			const chosen: InvidiousFormat = bestMp4 ?? bestAny;
+
+			// Infer MIME from container/type; Invidious does not always expose full codec/MIME details for YouTube progressive formats. – Invidious/YouTube issue
+			let mime =
+				chosen.type?.split(";")[0]?.trim() ||
+				(chosen.container === "webm"
+					? "video/webm"
+					: chosen.container === "mp4"
+					? "video/mp4"
+					: undefined);
+			// Always proxy via the instance (avoids CORS/signature churn) and carry itag if known.
+			const proxied = this.proxiedProgressive(host, id, chosen.itag);
+
+			// Known YouTube itag→MIME hints for progressive formats (historical values; may vary over time). – YouTube-specific
+			const ITAG_TO_MIME: Record<number, string> = {
+				18: "video/mp4", // 360p
+				22: "video/mp4", // 720p
+				59: "video/mp4", // 480p
+				43: "video/webm", // 360p (VP8/Vorbis)
+				44: "video/webm", // 480p
+				45: "video/webm", // 720p
+			};
```

Comment:
This is a constant it should go at the top level. This should also be a `Map` instead.